MIDI 44: Review CLIP ideas for MIDI data

JasiekKaczmarczyk opened the pull request:
[Review CLIP ideas for MIDI data](https://sacrebleu.atlassian.net/browse/MIDI-44)
[Embedding clustering evaluation](https://sacrebleu.atlassian.net/browse/MIDI-54)
[Streamlit dashboard to review the CLIP embeddings.](https://sacrebleu.atlassian.net/browse/MIDI-56)
[Beat random guess on validation](https://sacrebleu.atlassian.net/browse/MIDI-53)

I added quantization and dataset modules. I also implemented two encoder architectures one for pitch and the other for velocity and time data. Both are based on transformer architecture (more specifically BERT), but I think transformers are not particularly good for long sequences which we'll be dealing with, so maybe a convolutional architecture with attention layers in between (similar to our Unet) would be better. I could also try to implement this: https://arxiv.org/pdf/2302.10866.pdf

Commit:
MIDI-44: data + encoders

Commit:
MIDI-44: data + encoders

Commit:
MIDI-44: data quantization

JasiekKaczmarczyk commented:
I added prepare_dataset.py script for processing maestro dataset to quantized form. I used bin size of 7 for quantization of velocity, dstart and duration. Length of each sequence is 128.

Here is processed dataset: https://huggingface.co/datasets/JasiekKaczmarczyk/maestro-quantized

Commit:
MIDI-44: training loop

JasiekKaczmarczyk commented:
I implemented training loop and just like before I tried overfitting single batch of data.

Here are results:
![Zaznaczenie_006](https://github.com/Nospoko/midi-clip/assets/94491798/bf411d9e-d3e8-4e51-9754-4cbe5dc264f3)

To run training:
```sh
python train.py --config-name <config>
```

Next I will try to run model on whole dataset.

roszcz commented:
Nice! Can you check what the memory consumption looks like? How large can you make the batch on your machine?

JasiekKaczmarczyk commented:
> Nice! Can you check what the memory consumption looks like? How large can you make the batch on your machine?

I depends on the model size, but for the model from  ```config-default.yaml``` I used batch_size of 256 and it takes about 19% of memory.

roszcz commented:
We should be able estimate the batch size more accurately `batch_size = (gpu memory - model size) / sample size` - I'm wondering what's going to be the upper limit for a single gpu training. I can rent a RTX A6000 machine with 48GB - can you guesstimate if it's going to be enough to compare to what's described in the paper?

JasiekKaczmarczyk commented:
I tried to run training on full dataset for 10 epochs with batch size 512.

Here are results:
![Zaznaczenie_007](https://github.com/Nospoko/midi-clip/assets/94491798/a81396a6-67c7-4fe5-b2cc-311d20132492)

The accuracy score is very bad, but the model is learning. From what I can see from overfitted model, accuracy score starts getting pretty good when loss reaches about 1.0. I manged to reach about 5.5 loss during training on whole dataset, so I think the problem is with too small model for this task. Also the dataset might be too small.

Run on wandb: https://wandb.ai/jasiekkaczmarczyk/midi-clip/runs/916ecjl5

roszcz commented on data/prepare_dataset.py:
```diff
@@ -0,0 +1,90 @@
+
+    record = []
+
+    for subset in tqdm(piece_quantized.df.rolling(window=sequence_len, step=sequence_len)):
```
```suggestion
    for subset in tqdm(piece_quantized.df.rolling(window=sequence_len, step=sequence_len // 5)):
```

I think it's ok to have partially overlapping sequences, and the dataset will grow significantly.

It's definitely worth trying out - we could try decreasing the step for as long as the validation performance doesn't deteriorate

JasiekKaczmarczyk replied:
Yeah, that could work. I'll look into it.

roszcz commented:
@JasiekKaczmarczyk do you know what's up with those dips in the `train/loss` chart? Is this a result for a smaller batch size at the end of the epoch? If that's the case, would it make sense to normalize the loss with regard to batch size?

roszcz commented on data/prepare_dataset.py:
```diff
@@ -0,0 +1,90 @@
+
+import fortepyan as ff
+from tqdm import tqdm
+from quantizer import MidiQuantizer
```
```suggestion
from data.quantizer import MidiQuantizer
```

roszcz commented on data/prepare_dataset.py:
```diff
@@ -0,0 +1,90 @@
+    sequence_len = 128
+
+    # That's where I'm downloading the LTAFDB data
+    hf_dataset_path = "roszcz/maestro-v1"
```
```suggestion
    hf_dataset_path = "roszcz/maestro-v1-sustain"
```

This is a version of the dataset with the sustain pedal applied. This makes the data much closer to the original performance.

roszcz commented on data/prepare_dataset.py:
```diff
@@ -0,0 +1,90 @@
+    return record
+
+
+if __name__ == "__main__":
```
```suggestion
def main():
...



if __name__ == "__main__":
    main()
```

`__main__` shouldn't contain that much logic. The main plus here is that with this change I can run `prepare_dataset.main()` in ipython :)

roszcz commented on data/prepare_dataset.py:
```diff
@@ -0,0 +1,90 @@
+
+    record = []
+
+    for subset in tqdm(piece_quantized.df.rolling(window=sequence_len, step=sequence_len)):
```
```suggestion
    step = sequence_len // 2
    window_split = piece_quantized.df.rolling(window=sequence_len, step=step)
    for subset in tqdm(window_split, leave=False):
```

You can have nested progress bars with `leave=False`

roszcz commented on train.py:
```diff
@@ -0,0 +1,245 @@
+        test_ds = Subset(test_ds, indices=range(batch_size))
+
+    # dataloaders
+    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers)
```
```suggestion
    train_dataloader = DataLoader(train_ds, batch_size=batch_size, num_workers=num_workers, shuffle=True)
```

I think it's necessary to shuffle the train split - otherwise batches will be coming in creation order - mostly from the same piece and in sequence

roszcz commented on configs/config-default.yaml:
```diff
@@ -0,0 +1,43 @@
+
+models:
+  pitch_encoder:
+    num_embeddings: 128
```
Technically there's only 88 keys on a piano :)

Commit:
MIDI-44: updated train and data

Commit:
MIDI-44: updated train and data

Commit:
MIDI-44: added embedding clustering script

JasiekKaczmarczyk commented:
I added script to visualize embeddings of model. I used Kmeans to cluster pitch embeddings to 10 clusters and then applied PCA to reduce dimensions and check if clusters are in roughly the same place in velocity/time embedding space as in pitch embedding space.

Here are results on batch from train set:
![embeddings_clustering_train_set](https://github.com/Nospoko/midi-clip/assets/94491798/69a1200d-3042-410b-9659-fb99f32bf634)

Here are results on batch from validation set:
![embeddings_clustering_val_set](https://github.com/Nospoko/midi-clip/assets/94491798/de69ddbf-1b58-4a68-9a17-1cf7f3f4287c)

The model starts getting rough shape of the data distribution, but clearly is not trained yet. Both encoders have about 1 million parameters. My guess to improve results is to increase model size and batch size, so it has more samples to compare during one forward pass.

roszcz commented:
I think we should also try to check embedding consistency on other properties of the data - are the fragments in a cluster actually similar? For example, are Chopins close to Chopins in the embedding space? Are sonatas close to sonatas?

Commit:
MIDI-44: updated embedding clustering script

JasiekKaczmarczyk commented:
I added new function to ```embedding_clustering.py``` to visualize embeddings based on some query.

Here are some results:
![embeddings_query_chopin](https://github.com/Nospoko/midi-clip/assets/94491798/20975a0b-7ff2-494a-9643-b436366c167c)

![embeddings_query_liszt](https://github.com/Nospoko/midi-clip/assets/94491798/a93162e3-43cb-494c-8214-5d89a769f68a)

![embeddings_query_verdi](https://github.com/Nospoko/midi-clip/assets/94491798/d0ebde52-d481-470f-8e2a-9e6c212d9917)

roszcz commented:
Nice, looks promising 🚀

JasiekKaczmarczyk commented:
> Nice, looks promising 🚀

I think now main bottleneck is batch size. This model used batch size of 1024 and was trained for 10 epochs, dataset has 239k samples. This training technique relies heavily on seeing as much data in one pass as possible to make meaningful embeddings. Other solution is to train model longer, so each sample has more data for contrasting.

Also I think changing aggregation function for global embedding in encoders from ```torch.mean()``` to something else (maybe CLS token) might also help.

Commit:
MIDI-44: updated embedding clustering script

roszcz commented:
I think we need more ways to verify the quality of those embeddings, and I'd say the best thing to do is to start listening. We don't really have any meaningful way of checking if anything is improving other than looking at the loss value, which I think can be misleading

JasiekKaczmarczyk commented:
> 

I don't think that is very efficient. I could try implement it but I don't know if it would be informative, especially since velocity, dstart and duration are quantized.

roszcz committed:
cleanup validation

krianne committed:
dataset updates

- MidiDataset: move hf loading out of the data container
- Store source information in the quantized Maestro

JasiekKaczmarczyk committed:
Merge pull request #2 from Nospoko/MIDI-44/code-review

MIDI-44: Code Review

Commit:
MIDI-44: streamlit dashboard

JasiekKaczmarczyk commented:
I created streamlit dashboard to review CLIP embeddings. You can provide query for pitch embeddings and velocity/time embeddings separately. I used PCA to visualize embeddings. KDTree is used to find two closest embeddings, one in each set. Finally mp3 file is created in tmp directory and loaded into streamlit.

To run it:
```sh
streamlit run --server.port 4466 tests/streamlit_dashboard.py
```

roszcz commented:
Thanks!

To run this, I have to set up python path manually, e.g.:

```sh
PYTHONPATH=. streamlit run --server.port 4466 tests/streamlit_dashboard.py
```

Can you add support for the Maestro dataset? It is much higher quality than Giant Midi, so we should prioritize it for validation. I think you'll need to update you quantized clone to track the compose and piece title better - I did this in my code review PR, so hopefully it'll be low effort :)

Commit:
MIDI-44: updated streamlit dashboard

JasiekKaczmarczyk commented:
> Thanks!
> 
> To run this, I have to set up python path manually, e.g.:
> 
> ```shell
> PYTHONPATH=. streamlit run --server.port 4466 tests/streamlit_dashboard.py
> ```
> 
> Can you add support for the Maestro dataset? It is much higher quality than Giant Midi, so we should prioritize it for validation. I think you'll need to update you quantized clone to track the compose and piece title better - I did this in my code review PR, so hopefully it'll be low effort :)

Ok, I've updated it and used your maestro-quantized with source value for querying. Could you also make your model checkpoint visible on huggingface so I could use it for testing?

roszcz commented:
> Could you also make your model checkpoint visible on huggingface so I could use it for testing?

Here it is: https://huggingface.co/roszcz/midi-clip/blob/main/midi-clip-batch-256-2023-08-27-10-14.ckpt

And here's how it looks in wandb:

<img width="666" alt="image" src="https://github.com/Nospoko/midi-clip/assets/8056825/62882c58-f962-4255-bdbc-37d95c88b08f">
